File: University of Liverpool/Year 3/COMP 329 - Autonomous Mobile Robotics/lab3/deadreckoning_controller/deadreckoning_controller.py
# deadreckoning_controller.py

from controller import Supervisor
import pioneer_nav1 as pn
import math
from pioneer_nav1 import MoveState

def run_robot(robot):
    # Get the time step of the current world.
    timestep = int(robot.getBasicTimeStep())

    # Initialize the PioneerNavigation class
    nav = pn.PioneerNavigation(robot)
    time_elapsed = 0
    target_time = 0
    robot_velocity = 0.3
    
    # Define schedule
    schedule = [MoveState.CONFIGURE, MoveState.FORWARD, MoveState.ARC, MoveState.STOP]
    schedule_index = 0  # Start with the first action in the schedule
    
    display_action = ""

    # Initialize the configuration if the first action is CONFIGURE
    if schedule[0] == MoveState.CONFIGURE:
        nav.configure_initialise_parameters(2 * math.pi)

    # Main control loop
    while robot.step(timestep) != -1:
        current_state = schedule[schedule_index]

        if current_state == MoveState.CONFIGURE:
            # Check configuration parameters
            display_action = nav.configure_check_parameters(timestep)
            print(f"Action: {display_action} \tTrue Pose: {nav.get_real_pose()}")
            # Once configuration is done, move to the next action
            schedule_index = (schedule_index + 1) % len(schedule)
            target_time = 0
            time_elapsed = 0

        elif current_state == MoveState.FORWARD:
            # Move the robot forward for a specified distance
            if time_elapsed == 0:
                target_time = nav.forward(0.5, robot_velocity)  # Move 0.5 meters at 0.3 m/s
                display_action = "Forward Action: Moving forward"
            time_elapsed += timestep
            print(f"Action: {display_action} \tTrue Pose: {nav.get_real_pose()}")

            # Transition to the next action after moving forward for the target time
            if time_elapsed >= target_time:
                schedule_index = (schedule_index + 1) % len(schedule)
                target_time = 0
                time_elapsed = 0

        elif current_state == MoveState.ARC:
            # Move the robot in an arc
            if time_elapsed == 0:
                target_time = nav.arc(math.pi / 2, 0.7, robot_velocity)  # Arc 90 degrees at 0.3 rad/s
                display_action = "Arc Action: Moving in an arc"
            time_elapsed += timestep
            print(f"Action: {display_action} \tTrue Pose: {nav.get_real_pose()}")

            # Transition to the next action after the arc action completes
            if time_elapsed >= target_time:
                schedule_index = (schedule_index + 1) % len(schedule)
                target_time = 0
                time_elapsed = 0

        elif current_state == MoveState.STOP:
            # Stop the robot and end the loop
            nav.stop()
            display_action = "Stop Action: Robot stopped"
            print(f"Action: {display_action} \tTrue Pose: {nav.get_real_pose()}")
            break

if __name__ == "__main__":
    # Create the Supervisor instance
    my_robot = Supervisor()
    run_robot(my_robot)

File: University of Liverpool/Year 3/COMP 329 - Autonomous Mobile Robotics/lab3/DeadreckoningController/DeadreckoningController.java
// File: DeadreckoningController.java
// Description: Main controller for dead reckoning with a state machine.
// Author: Your Name

import com.cyberbotics.webots.controller.Supervisor;
import com.cyberbotics.webots.controller.Display;

public class DeadreckoningController {

    public static void main(String[] args) {

        Supervisor robot = new Supervisor();
        int timeStep = (int) Math.round(robot.getBasicTimeStep());

        PioneerNav1 nav = new PioneerNav1(robot);

        double time_elapsed = 0;
        double target_time = 0;
        double robot_velocity = 0.3;

        // Define schedule of actions
        PioneerNav1.MoveState[] schedule = { 
            PioneerNav1.MoveState.CONFIGURE, 
            PioneerNav1.MoveState.FORWARD, 
            PioneerNav1.MoveState.ARC, 
            PioneerNav1.MoveState.STOP 
        };
        int schedule_index = 0; // Start with the first action in the schedule

        String display_action = "";

        // Initialize the first action if it's CONFIGURE
        if (schedule[0] == PioneerNav1.MoveState.CONFIGURE) {
            nav.configure_initialise_parameters(2 * Math.PI);
            display_action = "Configuring...";
        }

        while (robot.step(timeStep) != -1) {
            PioneerNav1.MoveState state = nav.getState();

            // Handle different robot states
            if (state == PioneerNav1.MoveState.CONFIGURE) {
                display_action = nav.configure_check_parameters(timeStep);

                // If configuration is complete, move to next action in the schedule
                if (nav.getState() == PioneerNav1.MoveState.STOP) {
                    schedule_index = (schedule_index + 1) % schedule.length;
                    state = schedule[schedule_index];
                    nav.setState(state); // Manually set the state for the next action
                }

            } else if (state == PioneerNav1.MoveState.FORWARD) {
                target_time = nav.forward(0.5, robot_velocity); // Move forward 0.5 meters
                display_action = "Moving Forward 0.5m";

                // Check if the forward action duration has passed
                if (time_elapsed >= target_time) {
                    nav.stop();
                    time_elapsed = 0;
                    schedule_index = (schedule_index + 1) % schedule.length;
                    state = schedule[schedule_index];
                    nav.setState(state);
                }

            } else if (state == PioneerNav1.MoveState.ARC) {
                target_time = nav.arc(Math.PI / 2, 0.7, robot_velocity); // Arc around ICR 0.7m away
                display_action = "Arcing around ICR";

                // Check if the arc action duration has passed
                if (time_elapsed >= target_time) {
                    nav.stop();
                    time_elapsed = 0;
                    schedule_index = (schedule_index + 1) % schedule.length;
                    state = schedule[schedule_index];
                    nav.setState(state);
                }

            } else if (state == PioneerNav1.MoveState.STOP) {
                nav.stop();
                display_action = "Stopped";

                // After stopping, cycle back to the first action or end the program
                schedule_index = 0;
                nav.setState(schedule[schedule_index]);
            }

            // Increment elapsed time only if not in CONFIGURE state
            if (state != PioneerNav1.MoveState.CONFIGURE) {
                time_elapsed += timeStep;
            }

            // Output current action and true pose
            Pose true_pose = nav.get_real_pose();
            System.out.println("Action: " + display_action + " \t" + "True Pose: " + true_pose);
        }
    }
}
